JuliaStream.jl/src/oneAPIStream.jl

```diff
@@ -0,0 +1,170 @@
+using Base.Iterators: println
+using Base.Iterators: println
+using Printf: Iterators
+
+include("Stream.jl")
+using oneAPI
+
+const oneData = StreamData{T,oneArray{T}} where {T}
+const DotWGSize = 256::Int
+
+function devices()::Vector{DeviceWithRepr}
+  all = map(oneL0.devices, oneL0.drivers()) |> Iterators.flatten |> Iterators.collect
+  map(dev -> (dev, repr("text/plain", dev), "oneAPi.jl"), all)
+end
+
+function make_stream(
+  arraysize::Int,
+  scalar::T,
+  device::DeviceWithRepr,
+  silent::Bool,
+)::Tuple{oneData{T},Int} where {T}
+
+  oneAPI.allowscalar(false)
+  oneAPI.device!(device[1])
+
+  props = oneL0.compute_properties(oneAPI.device())
+  groupsize = min(props.maxTotalGroupSize, arraysize)
+
+  if arraysize % groupsize != 0
+    error("arraysize ($(arraysize)) must be divisible by $(groupsize)!")
+  end
+
+  if !silent
+    println("Using L0 device: $(repr("text/plain",device[1]))")
+    println("Kernel parameters   : <<<$(arraysize),$(groupsize)>>>")
+  end
+  return (
+    oneData{T}(
+      oneArray{T}(undef, arraysize),
+      oneArray{T}(undef, arraysize),
+      oneArray{T}(undef, arraysize),
+      scalar,
+      arraysize,
+    ),
+    groupsize,
+  )
+end
+
+function init_arrays!(data::oneData{T}, _, init::Tuple{T,T,T}) where {T}
+  oneAPI.fill!(data.a, init[1])
+  oneAPI.fill!(data.b, init[2])
+  oneAPI.fill!(data.c, init[3])
+end
+
+function copy!(data::oneData{T}, groupsize::Int) where {T}
+  function kernel(a, c)
+    i = get_global_id()
+    @inbounds c[i] = a[i]
+    return
+  end
+  @oneapi items = groupsize groups = data.size ÷ groupsize kernel( # 
+    data.a,
+    data.c,
+  )
+  oneAPI.synchronize()
+end
+
+function mul!(data::oneData{T}, groupsize::Int) where {T}
+  function kernel(b, c, scalar)
+    i = get_global_id()
+    @inbounds b[i] = scalar * c[i]
+    return
+  end
+  @oneapi items = groupsize groups = data.size ÷ groupsize kernel( #
+    data.b,
+    data.c,
+    data.scalar,
+  )
+  oneAPI.synchronize()
+end
+
+function add!(data::oneData{T}, groupsize::Int) where {T}
+  function kernel(a, b, c)
```

Comment:
It's a very minor detail, but you might want to put the `T` typevar on the inner kernel function instead, e.g., by using `a::AbstractArray{T}, ...` as arguments. That way you enforce specialization of the kernel method, whereas the outer CPU method doesn't necessarily need to be specialized (and you can use `eltype(data)` if you need the element type there).

But these things only matter when you work with very complicated inputs.

Reply:
Does this apply to all other GPU implementations or just oneAPI? 
Ideally this would be an exemplar for writing kernels so would you recommend I annotate every other GPU implementation like this?
 
Aside:
I thought about doing this with more precise types (e.g `::CuDeviceVector{T}` for CUDA) but felt like the actual type is an implementation detail for the kernel launch macro  (`@cuda`, `@oneapi`, etc) so I guess `AbstractArray{T}` is the best we an do.

Reply:
This applies to all Julia code, not just the oneAPI GPU implementation. If you add a typevar (`::T .. .where T`) you force the compiler to specialize. That's needed for the GPU kernel, because we don't want or support dynamic code for GPUs, but the host function (here, `add!`) may remain underspecialized if the compiler thinks that's useful.

And yeah, the `AbstractArray` suggestion is so that you don't have to deal with the exact device-side types, which are an implementation detail.